File: packages/app/src/desktop/notifications/desktop-notifications.ts
import { invokeDesktopCommand } from "@/desktop/tauri/invoke-desktop-command";
import { getTauri } from "@/utils/tauri";

export const DESKTOP_NOTIFICATION_CLICK_EVENT = "desktop-notification-click";

export interface DesktopNotificationInput {
  title: string;
  body?: string;
  data?: Record<string, unknown>;
}

export interface DesktopNotificationClickPayload {
  data?: Record<string, unknown>;
}

export type DesktopNotificationClickHandler = (
  payload: DesktopNotificationClickPayload
) => void;

export type DesktopNotificationClickUnlisten = () => void;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export async function sendDesktopNotification(
  input: DesktopNotificationInput
): Promise<boolean> {
  try {
    await invokeDesktopCommand("send_desktop_notification", { input });
    return true;
  } catch (error) {
    console.warn("[OSNotifications][Desktop] Failed to send desktop notification", error);
    return false;
  }
}

export async function listenToDesktopNotificationClicks(
  handler: DesktopNotificationClickHandler
): Promise<DesktopNotificationClickUnlisten> {
  const listen = getTauri()?.event?.listen;
  if (typeof listen !== "function") {
    throw new Error("Tauri event API is unavailable.");
  }

  const unlisten = await listen(DESKTOP_NOTIFICATION_CLICK_EVENT, (event: unknown) => {
    const payload = isRecord(event) && isRecord(event.payload) ? event.payload : null;
    if (!payload) {
      return;
    }

    handler({
      data: isRecord(payload.data) ? payload.data : undefined,
    });
  });

  return typeof unlisten === "function" ? unlisten : () => {};
}

File: packages/desktop/src-tauri/src/desktop_notifications.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tauri::{AppHandle, Runtime};

pub const DESKTOP_NOTIFICATION_CLICK_EVENT: &str = "desktop-notification-click";

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DesktopNotificationInput {
    pub title: String,
    pub body: Option<String>,
    pub data: Option<Value>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct DesktopNotificationClickPayload {
    data: Option<Value>,
}

#[cfg(target_os = "macos")]
mod platform {
    use super::{
        AppHandle, DesktopNotificationClickPayload, DesktopNotificationInput,
        DESKTOP_NOTIFICATION_CLICK_EVENT,
    };
    use std::sync::OnceLock;
    use tauri::{async_runtime, Emitter, Runtime};

    const MACOS_NOTIFICATION_SENDER_BUNDLE_ID: &str = "dev.paseo.desktop";
    static MACOS_NOTIFICATION_SENDER_INIT: OnceLock<Result<(), String>> = OnceLock::new();

    fn ensure_notification_sender() -> Result<(), String> {
        MACOS_NOTIFICATION_SENDER_INIT
            .get_or_init(|| {
                mac_notification_sys::set_application(MACOS_NOTIFICATION_SENDER_BUNDLE_ID)
                    .map_err(|error| {
                        format!(
                            "Failed to initialize macOS notification sender for {}: {error}",
                            MACOS_NOTIFICATION_SENDER_BUNDLE_ID
                        )
                    })
            })
            .clone()
    }

    pub fn send_notification<R: Runtime>(
        app: AppHandle<R>,
        input: DesktopNotificationInput,
    ) -> Result<(), String> {
        let _task = async_runtime::spawn_blocking(move || {
            if let Err(error) = ensure_notification_sender() {
                log::warn!("{error}");
                return;
            }

            let mut notification = mac_notification_sys::Notification::default();
            notification
                .title(&input.title)
                .message(input.body.as_deref().unwrap_or(""))
                .wait_for_click(true);

            match notification.send() {
                Ok(mac_notification_sys::NotificationResponse::Click) => {
                    let payload = DesktopNotificationClickPayload { data: input.data };
                    if let Err(error) = app.emit(DESKTOP_NOTIFICATION_CLICK_EVENT, payload) {
                        log::warn!("failed to emit desktop notification click event: {error}");
                    }
                }
                Ok(_) => {}
                Err(error) => {
                    log::warn!("failed to send macOS notification: {error}");
                }
            }
        });

        Ok(())
    }
}

#[cfg(not(target_os = "macos"))]
mod platform {
    use super::{AppHandle, DesktopNotificationInput};
    use tauri::Runtime;
    use tauri_plugin_notification::NotificationExt;

    pub fn send_notification<R: Runtime>(
        app: AppHandle<R>,
        input: DesktopNotificationInput,
    ) -> Result<(), String> {
        let mut notification = app.notification().builder().title(&input.title);
        if let Some(body) = &input.body {
            notification = notification.body(body);
        }

        notification
            .show()
            .map_err(|error| format!("Failed to send desktop notification: {error}"))?;

        Ok(())
    }
}

#[tauri::command]
pub async fn send_desktop_notification<R: Runtime>(
    app: AppHandle<R>,
    input: DesktopNotificationInput,
) -> Result<(), String> {
    platform::send_notification(app, input)
}
